cardano-node/src/Cardano/Tracing/Tracers.hs

```diff
@@ -634,6 +641,83 @@ mkConsensusTracers mbEKGDirect trSel verb tr nodeKern fStats = do
      when (isRollForward ev) $
        sendEKGDirectCounter ekgDirect "cardano.node.metrics.served.header.counter.int"
 
+traceBlockFetchClientMetrics
+  :: forall blk remotePeer.
+     ( )
+  => Maybe EKGDirect
+  -> STM.TVar (SMap.Map SlotNo NominalDiffTime)
+  -> STM.TVar Int64
+  -> STM.TVar Int64
+  -> STM.TVar Int64
+  -> Tracer IO (TraceLabelPeer remotePeer (TraceFetchClientState (Header blk)))
+  -> Tracer IO (TraceLabelPeer remotePeer (TraceFetchClientState (Header blk)))
+traceBlockFetchClientMetrics Nothing _ _ _ _ tracer = tracer
+traceBlockFetchClientMetrics (Just ekgDirect) slotMapVar cdf1sVar cdf2sVar cdf5sVar tracer = Tracer bfTracer
+
+  where
+
+    -- We track the cdf for 1s, 2s and 5s.
+    updateCDFCounter :: NominalDiffTime -> Int -> Int64 -> STM (Double, Double, Double)
+    updateCDFCounter delay size step = do
+
+      when (delay < 1) $ do
+        STM.modifyTVar' cdf1sVar (+ step)
+      when (delay < 2) $ do
+        STM.modifyTVar' cdf2sVar (+ step)
+      when (delay < 5) $ do
+        STM.modifyTVar' cdf5sVar (+ step)
+
+      cdf1s <- STM.readTVar cdf1sVar
+      cdf2s <- STM.readTVar cdf2sVar
+      cdf5s <- STM.readTVar cdf5sVar
+      return (fromIntegral cdf1s / fromIntegral size, fromIntegral cdf2s / fromIntegral size,
+              fromIntegral cdf5s / fromIntegral size)
```

Comment:
I am assuming that the overall fraction of blocks received is tracked elsewhere - i.e. here it is a fraction, somewhere else we know the total population size...

Reply:
That is tracked in `slotMapVar` which contains the `k/2` latest blocks.